---
title: "Gully eDNA Biological Data Analysis"
author: "Shannon Myles"
date: '2021-11-24'
output:
  html_document: default
  pdf_document: default
---
```{r, echo=FALSE}
knitr::opts_chunk$set(error = FALSE)
```

```{r, Installation of required packages, results=FALSE, message=FALSE, warning=FALSE}
# Run lines as needed depending on what you have already installed on your R machine
#install.packages("remotes")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#remotes::install_github("jbisanz/qiime2R")
#BiocManager::install("phyloseq")
#install.packages("tidyverse")
#install.packages("ggplot2")
#BiocManager::install("microbiome")
#install.packages("vegan")
#install.packages("devtools")
#devtools::install_github("adw96/breakaway")
#remotes::install_github("cmmr/rbiom")
#devtools::install_github("joey711/phyloseq")
#install.packages("ggthemes")

# Loading feature tables in .qza as subsettable objects 
library(qiime2R)            # from github: jbisanz
# Loading experimental data - a comprehensive wrapper for processing, analysis, and plotting
library(phyloseq)
# These two accompany phyloseq (both biom and phyloseq - from github: joey711)
library(biom)
library(rbiom)                  # from Github: cmmr
#Dataframe management, plotting
library(tidyverse)
library(ggplot2)
library(ggthemes)
#Transformation and summary function
library(microbiome)
#Species richness
#library(breakaway)
# Statistical testing for diversity diifferences*
library(vegan)

# set working directory to access files
setwd("C:/Users/18192/Documents/DAL/Honours/Qiime2/Final_data")
```

# Creating helper functions necessary for the analysis
```{r functions derived from or directly from Djurhuus et al. 2020 & Kelly et al. 2019: for splitting species names, adding taxonomy to highest level available, agglomerating into unique taxa at a desired level, and the eDNA index}
# this is a function that Djurhuus references from https://rdrr.io/github/mworkentine/mattsUtils/src/R/microbiome_helpers.R#sym-add_taxonomy_column
split_species = function(string, n = 2) {
  splits = str_split(string, " ", n + 1)
  res = map_if(splits, ~length(.x) > 2, ~.x[1:n]) %>%
    map_chr(str_c, collapse = " ")
  return(res)
}

# this is function that Djurhuus references from https://rdrr.io/github/mworkentine/mattsUtils/src/R/microbiome_helpers.R#sym-add_taxonomy_column
add_taxonomy_column = function(physeq, num_species = 1) {
  tax_df = as.data.frame(tax_table(physeq)@.Data) %>%
    rownames_to_column("OTU") %>%
    mutate(specie = split_species(specie, n = num_species)) %>%
    mutate(Taxonomy =
      case_when(
        is.na(class)  ~ str_c("", phyla),
        is.na(order)  ~ str_c("", class),
        is.na(family)  ~ str_c("", order),
        is.na(genus)   ~ str_c("", family),
        is.na(specie) ~ str_c("", genus),
        TRUE ~ str_c(genus, " ", specie)
      )
    )

  tax = as.matrix(tax_df[, -1])
  rownames(tax) = tax_df$OTU
  tax_table(physeq) = tax_table(tax)

  return(physeq)
}

# Custom tax_agglomerate so that the level at which to agglomerate becomes first tax column
# This is because otherwise, conflicting tax IDs at higher level, for the same tax ID at the level to agglomerate yield two (or n) agglomerations as output - but you really just want 1. 
# provided by GitHub user Adriaaula as a response to the issue opened by Anni Djurhuus here: https://github.com/joey711/phyloseq/issues/941
tax_glom_custom = function(physeq, tax.col){
  
  require(tidyverse)
  
  reorderedtax = data.frame(tax_table(physeq)) %>% 
    # Puts the selected column first with everything else aftwds
    dplyr::select(one_of(tax.col), everything()) %>% 
    as.matrix() %>%                                                    # My addition
    tax_table()
  
  tax_table(physeq) = reorderedtax
  
  return(tax_glom(physeq, taxrank = tax.col))
}


# This comes from GitHub user (Phyloseq maintainer) Joey711
pop_taxa = function(physeq, badTaxa){
  allTaxa = taxa_names(physeq)
  myTaxa <- allTaxa[!(allTaxa %in% badTaxa)]
  return(prune_taxa(myTaxa, physeq))
}

# eDNA index directly from Kelly et al. 2019
# math defined below
eDNA_index <- function(x) { #where x is a dataframe with taxa/OTUs/etc in rows, and samples in columns
  rowMax <- function(x){apply(x, MARGIN = 1, FUN = max)}
  temp <- sweep(x, MARGIN = 2, STATS = colSums(x), FUN = "/") #create proportion for each taxon/OTU within each sample
  sweep(temp, MARGIN = 1, STATS = rowMax(temp), FUN = "/")
}
```

# metadata pre-processing
```{r, metadata pre-processing, results=FALSE, message=FALSE}
# 12S
# import and make sample id column the row names
S12_metadata <-  read_tsv("./12S_exports/metadata12S.tsv") %>%
  column_to_rownames(var = "sampleid")
# add units to columns and effectively renders data type character instead of numeric
S12_metadata$size_fraction <- paste(as.character(S12_metadata$size_fraction), 'µm')
S12_metadata <- S12_metadata %>% 
  mutate(disc_depth = paste(as.character(S12_metadata$depth), 'm')) %>% 
  mutate(timeofyear = paste(as.character(S12_metadata$year), S12_metadata$season))

# 16S
# import and make sample id column the row names
S16_metadata <-  read_tsv("./16S_exports/metadata.tsv") %>%
  column_to_rownames(var = "sampleid")
# add units to columns and effectively renders data type character instead of numeric
S16_metadata$size_fraction <- paste(as.character(S16_metadata$size_fraction), 'µm')
S16_metadata <- S16_metadata %>%
  mutate(disc_depth = paste(as.character(S16_metadata$depth), 'm')) %>% 
  mutate(timeofyear = paste(as.character(S16_metadata$year), S16_metadata$season))

# 18S
# import and make sample id column the row names
S18_metadata <-  read_tsv("./18S_exports/metadata.tsv") %>%
  column_to_rownames(var = "sampleid")
# add units to columns and effectively renders data type character instead of numeric
S18_metadata$size_fraction <- paste(as.character(S18_metadata$size_fraction), 'µm')
S18_metadata <- S18_metadata %>%
  mutate(disc_depth = paste(as.character(S18_metadata$depth), 'm')) %>% 
  mutate(timeofyear = paste(as.character(S18_metadata$year), S18_metadata$season))

```

# Taxonomic tables pre-processing
```{r, Taxonomic tables pre-processing, results = 'hide', message=FALSE}
# Reading in tax_tables
S12_tax <- read_tsv("./12S_exports/taxa_FishCard/taxonomy.tsv") %>%
  remove_rownames %>%                          # clear row names
  column_to_rownames(var="Feature ID") %>%     # make Feature ID the new row names
  as.matrix()                                  # as.matrix because that's the format needed for phyloseq

# Reading in tax_tables and doing some string parsing to remove D_0__ taxonomic level suffixes
S16_tax <- read_tsv("./16S_exports/taxa/taxonomy.tsv") %>%
      mutate_at("domain", str_replace, "D_0__", "")%>%
      mutate_at("phyla", str_replace, "D_1__", "")%>%
      mutate_at("class", str_replace, "D_2__", "")%>%
      mutate_at("order", str_replace, "D_3__", "")%>%
      mutate_at("family", str_replace, "D_4__", "")%>%
      mutate_at("genus", str_replace, "D_5__", "")%>%
      mutate_at("specie", str_replace, "D_6__", "")%>% 
      remove_rownames %>%
      column_to_rownames(var="Feature ID") %>%
      as.matrix()

# Reading in tax_tables and doing some string parsing to remove d__ taxonomic level suffixes
S18_tax <- read_tsv("./18S_exports/taxa_final/taxonomy.tsv") %>%
      mutate_at("domain", str_replace, "d__", "")%>%
      mutate_at("phyla", str_replace, "p__", "")%>%
      mutate_at("class", str_replace, "c__", "")%>%
      mutate_at("order", str_replace, "o__", "")%>%
      mutate_at("family", str_replace, "f__", "")%>%
      mutate_at("genus", str_replace, "g__", "")%>%
      mutate_at("specie", str_replace, "s__", "")%>% 
      remove_rownames %>%
      column_to_rownames(var="Feature ID") %>%
      as.matrix()
```
# Loading in the data into phyloseq objects to merge them into 1 phyloseq object per marker region
```{r, Loading in various data sets for each marker further (minimal) pre-processing & merging into 1 global phyloseq object per marker, warning=FALSE}
# Appropriate sample names for 12S and 16S (there's slight errors in the samples names that do not match between all markers' samples names)
# Here i am manually naming them identically to 18S sample names
S12_samples_names <- c("2014F-GULD04-100mL", "2014F-GULD04-1mL", "2014S-GULD04-250mL", "2016F-GULD4-100m-L", "2016F-GULD4-250m-L", "2016S-GULD3-40m-L", "2017F-GULD04-1mL", "2017S-GULD04-60mL", "2018F-SG23-1mL", "2018F-SG23-20mL", "2018F-SG23-250mL", "2018F-SG23-60mL", "2018S-GULD04-250mL", "2019S-GULD04-1mL", "2019S-GULD04-250mL")

S16_samples_names <- c("2014F-GULD04-100mL", "2014F-GULD04-100mS", "2014F-GULD04-1mL",   "2014F-GULD04-1mS",   
                       "2014F-GULD04-20mL", "2014F-GULD04-20mS","2014F-GULD04-250mL", "2014F-GULD04-250mS", 
                       "2014S-GULD04-100mL", "2014S-GULD04-100mL" , "2014S-GULD04-1mL", "2014S-GULD04-1mS" ,  
                       "2014S-GULD04-20mL" , "2014S-GULD04-20mS" , "2014S-GULD04-250mL","2014S-GULD04-250mS", 
                       "2016F-GULD4-100m-L",  "2016F-GULD4-100m-S" , "2016F-GULD4-1m-L"   , "2016F-GULD4-1m-S" ,  
                       "2016F-GULD4-20m-L"  , "2016F-GULD4-20m-S" ,  "2016F-GULD4-250m-L" , "2016F-GULD4-250m-S" , 
                       "2016S-GULD3-1m-L"   ,"2016S-GULD3-1m-S"  ,  "2016S-GULD3-20m-L"  , "2016S-GULD3-20m-S"  , 
                       "2016S-GULD3-40m-L" ,  "2016S-GULD3-40m-S"  ,"2016S-GULD3-80m-L","2016S-GULD3-80m-S"  , 
                       "2017F-GULD04-1mL"   , "2017F-GULD04-1mS"  ,  "2017F-GULD04-20mL"  ,"2017F-GULD04-20mS" ,  
                       "2017F-GULD04-250mL" , "2017F-GULD04-250mS" , "2017F-GULD04-60mL"  , "2017F-GULD04-60mS"  ,
                       "2017S-GULD04-1mL"  ,  "2017S-GULD04-1mS"   , "2017S-GULD04-20mL" ,  "2017S-GULD04-20mS" ,  
                       "2017S-GULD04-250mL" ,"2017S-GULD04-250mS"  ,"2017S-GULD04-60mL"  , "2017S-GULD04-60mS"  , 
                       "2018F-SG23-1mL"    ,  "2018F-SG23-1mS"     ,"2018F-SG23-20mL"    , "2018F-SG23-20mS"    , 
                       "2018F-SG23-250mL"   , "2018F-SG23-250mS"  ,  "2018F-SG23-60mL", "2018F-SG23-60mS"    , 
                       "2018S-GULD04-1mS"  ,  "2018S-GULD04-20mL" ,  "2018S-GULD04-20mS" ,  "2018S-GULD04-250mL" ,
                       "2018S-GULD04-250mS" , "2018S-GULD04-60mL"  , "2018S-GULD04-60mS" ,  "2019S-GULD04-1mS"  ,  
                       "2019S-GULD04-20mL"  ,"2019S-GULD04-20mS"  , "2019S-GULD04-250mL" , "2019S-GULD04-250mS",  
                       "2019S-GULD04-60mL",   "2019S-GULD04-60mS" )

# 12S marker
S12_ASVs <- read_qza("./12S_exports/dada2_output/dada2_table_filt.qza")[["data"]] %>%
  as.data.frame() 
names(S12_ASVs) <- S12_samples_names         # Assign new names 
S12_ASVs <- otu_table(S12_ASVs, taxa_are_rows = TRUE)         # reading as phyloseq

S12_tax <- tax_table(S12_tax)                                 # reading as phyloseq
S12_metadata <- sample_data(S12_metadata)                     # reading as phyloseq

S12_tree <- read_qza("./12S_exports/asvs-tree.qza")[["data"]] %>%
  read_tree(taxa_are_rows = TRUE)                             # reading as phyloseq


# 16S marker
S16_ASVs <- read_qza("./16S_exports/dada2_output/dada2_table_final.qza")[["data"]] %>%
  as.data.frame() 
names(S16_ASVs) <- S16_samples_names        # Assign new names 
S16_ASVs <- otu_table(S16_ASVs, taxa_are_rows = TRUE)         # reading as phyloseq

S16_tax <- tax_table(S16_tax)                                 # reading as phyloseq
S16_metadata <- sample_data(S16_metadata)                     # reading as phyloseq

S16_tree <- read_qza("./16S_exports/asvs-tree.qza")[["data"]] %>%
  read_tree(taxa_are_rows = TRUE)                             # reading as phyloseq


# 18S marker
S18_ASVs <- read_qza("./18S_exports/dada2_output/dada2_table_filt.qza")[["data"]] %>%
  as.data.frame() %>%
  otu_table(taxa_are_rows = TRUE)                             # reading as phyloseq

S18_tax <- tax_table(S18_tax)                                 # reading as phyloseq
S18_metadata <- sample_data(S18_metadata)                     # reading as phyloseq

S18_tree <- read_qza("./18S_exports/asvs-tree.qza")[["data"]] %>%
  read_tree(taxa_are_rows = TRUE)                             # reading as phyloseq

# Merging phyloseq data into 1 phyloseq object per marker
(S12_physeq <-  merge_phyloseq(S12_ASVs, S12_tax, S12_metadata, S12_tree))
(S16_physeq <-  merge_phyloseq(S16_ASVs, S16_tax, S16_metadata, S16_tree))
(S18_physeq <-  merge_phyloseq(S18_ASVs, S18_tax, S18_metadata, S18_tree))
```
# Agglomerating data at family taxonomic level
```{r, Agglomeration functions to merge all ASVs at x tax level and sums up their read counts, warning=FALSE}

# Agglomeration functions to merge all ASVs at x tax level and sums up their read counts
S12_family <- tax_glom(S12_physeq, "family")
S16_family <- tax_glom(S16_physeq, "family")
S18_family <- tax_glom(S18_physeq, "family")

```


```{r Subset only taxa with a family annotation, assign the family name to new column: Taxonomy, make this column the taxa_names(), warning = FALSE}
S12_family <- add_taxonomy_column(S12_family) %>% 
  subset_taxa(Taxonomy !="NA") %>% 
  tax_glom_custom("family")
taxa_names(S12_family) <- paste(S12_family@tax_table[,1])

S16_family <- add_taxonomy_column(S16_family) %>% 
  subset_taxa(Taxonomy !="NA") %>% 
  tax_glom_custom("family")
taxa_names(S16_family) <- paste(S16_family@tax_table[,1])

S18_family <- add_taxonomy_column(S18_family) %>% 
  subset_taxa(Taxonomy !="NA") %>% 
  tax_glom_custom("family")
taxa_names(S18_family) <- paste(S18_family@tax_table[,1])
```

# Clean contaminants and rename ambiguous taxa
```{r}
S12_bad <- c("Suidae", "Hominidae", "Bovidae", "Canidae")
S12_family <- pop_taxa(S12_family, S12_bad)

S16_bad <- c("Mitochondria")
S16_family <- pop_taxa(S16_family, S16_bad)

S18_bad <- c("uncultured", "Incertae_Sedis")
S18_family <- pop_taxa(S18_family, S18_bad)
```

# Extract a vector of the sum of read counts for each family (top 10 families for 16 and 18S)
```{r, Extract a vector the sum of of read counts for each family}
S12_fam_read_counts <- rowSums(otu_table(S12_family)) %>% 
  sort(decreasing = TRUE)

S16_fam_read_counts <- rowSums(otu_table(S16_family)) %>% 
  sort(decreasing = TRUE)
S16_fam_read_counts_topn <- S16_fam_read_counts[1:10]

S18_fam_read_counts <- rowSums(otu_table(S18_family)) %>% 
  sort(decreasing = TRUE)
S18_fam_read_counts_topn <- S18_fam_read_counts[1:10]
```

# Create a subset of the bigger picture, only including top n taxa defined above
```{r}
topn_S16_families <- prune_taxa(names(S16_fam_read_counts_topn), S16_family)
# Specifying some ambiguous taxonomy
S16_revised_taxa_names <- c("Cyanobiaceae", 
                            "Pseudoalteromonadaceae", 
                            "Alteromonadaceae", 
                            "Thioglobaceae", 
                            "Flavobacteria: NS9 marine group", 
                            "Flavobacteriaceae",      
                            "Sphingomonadaceae",      
                            "Rhodobacteraceae",       
                            "Pelagibacterales: Clade II",               
                            "Pelagibacterales: Clade I")

# Re-specifying for phyloseq object
taxa_names(topn_S16_families) <- S16_revised_taxa_names

obj <- data.frame(as(tax_table(topn_S16_families), "matrix"))
obj$family <- S16_revised_taxa_names
tax_table(topn_S16_families) <- obj %>% as.matrix() %>% tax_table()

# re-specifying for vector - plotting
names(S16_fam_read_counts_topn) <- S16_revised_taxa_names


topn_S18_families <- prune_taxa(names(S18_fam_read_counts_topn), S18_family)
```


# Creating ordered barplots of family read counts
```{r, Creating 12S families barplot}
ggplot() + coord_flip() + geom_bar(aes(x= reorder(names(S12_fam_read_counts), S12_fam_read_counts), y=S12_fam_read_counts), stat='identity', color = "#4169E1", fill = "#4169E1") + xlab("") + ylab("Read counts across all experiments (n=15)")
ggsave("12S families read counts barplot.png")
```

```{r, Creating 12S families barplot}
ggplot() + coord_flip() + geom_bar(aes(x= reorder(names(S16_fam_read_counts_topn), S16_fam_read_counts_topn), y=S16_fam_read_counts_topn), stat='identity',  fill = "#4169E1") + xlab("") + ylab("Read counts across all experiments (n=70)")
ggsave("16S families read counts barplot.png")
```


```{r, Creating 12S families barplot}
ggplot() + coord_flip() + geom_bar(aes(x= reorder(names(S18_fam_read_counts_topn), S18_fam_read_counts_topn), y=S18_fam_read_counts_topn), stat='identity', fill = "#4169E1") + xlab("") + ylab("Read counts across all experiments (n=69)")
ggsave("18S families read counts barplot.png")
```

# Standardize datasets with the eDNA index
```{r}
# computing eDNA index for 12S data
S12_family_eDNA_otu <- as.data.frame(otu_table(S12_family)) %>%
  eDNA_index() %>% 
  otu_table(taxa_are_rows = TRUE)
# Prune out taxa who dont have any data in any sample
S12_family_eDNA_otu = prune_taxa(taxa_sums(S12_family_eDNA_otu)>0, S12_family_eDNA_otu)
# Re-create a standardized phyloseq object
(S12_family_eDNA_indexed <- merge_phyloseq(S12_family_eDNA_otu,
                                 tax_table(S12_family),
                                 sample_data(S12_family)
                                 ))

# computing eDNA index for 16S data
S16_family_eDNA_otu <- as.data.frame(otu_table(topn_S16_families)) %>%
  eDNA_index() %>% 
  otu_table(taxa_are_rows = TRUE)
# Prune out taxa who dont have any data in any sample
S16_family_eDNA_otu = prune_taxa(taxa_sums(S16_family_eDNA_otu)>0, S16_family_eDNA_otu)
# Re-create a standardized phyloseq object
(S16_family_eDNA_indexed <- merge_phyloseq(S16_family_eDNA_otu,
                                 tax_table(topn_S16_families),
                                 sample_data(topn_S16_families)
                                 ))

# computing eDNA index for 18S data
S18_family_eDNA_otu <- as.data.frame(otu_table(topn_S18_families)) %>%
  eDNA_index() %>% 
  otu_table(taxa_are_rows = TRUE)
# Prune out taxa who dont have any data in any sample
S18_family_eDNA_otu = prune_taxa(taxa_sums(S18_family_eDNA_otu)>0, S18_family_eDNA_otu)
# Re-create a standardized phyloseq object
(S18_family_eDNA_indexed <- merge_phyloseq(S18_family_eDNA_otu,
                                 tax_table(topn_S18_families),
                                 sample_data(topn_S18_families)
                                 ))
```

```{r}
p <- plot_bar(S12_family_eDNA_indexed, x = "timeofyear", fill = "family")
p + geom_bar(aes(color=family, fill = family), stat="identity", position="stack") + labs(x = "Sampling event)", y = "Abundance (eDNA indexed)")
ggsave("12S families eDNA index over time.png")
```

```{r}
p <- plot_bar(S16_family_eDNA_indexed, x = "timeofyear", fill = "family")
p + geom_bar(aes(color=family, fill = family), stat="identity", position="stack") + labs(x = "Sampling event)", y = "Abundance (eDNA indexed)")
ggsave("16S families eDNA index over time.png")
```

```{r}
p <- plot_bar(S18_family_eDNA_indexed, x = "timeofyear", fill = "family")
p + geom_bar(aes(color=family, fill = family), stat="identity", position="stack") + labs(x = "Sampling event)", y = "Abundance (eDNA indexed)")
ggsave("18S families eDNA index over time.png")
```



