---
title: "Gully_marker_merger"
author: "Shannon Myles"
date: "2021-12-01"
output: html_notebook
---
# Installing packages as needed & Loading them for the job
```{r, message=FALSE}
#install.packages("remotes")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#remotes::install_github("jbisanz/qiime2R")
#remotes::install_github("joey711/biom")
#BiocManager::install("phyloseq")
#install.packages("tidyverse")
#install.packages("ggplot2")


library(phyloseq)     # A data structure / landing place for seq data

# The following interact very well with Phyloseq
library(qiime2R)      # Importing QIIME artifacts
library(biom)         # Handling BIOM tables
library(rbiom)        # More handling BIOM tables
library(ape)          # Phylogeny handler

# The more typical toolbox
library(tidyverse)    # Data handling and so much more
library(reshape2)     # Statistical tool
library(ggplot2)      # Plotting library
library(matrixStats)  # More stats

# Setting working directory to access files
setwd("C:/Users/18192/Documents/DAL/Honours/Qiime2/Final_data")
```

# Creating custom functions necessary for the analysis
```{r functions derived from or directly from Djurhuus et al. 2020 & Kelly et al. 2019: for splitting species names, adding taxonomy to highest level available, agglomerating into unique taxa at a desired level, and the eDNA index}
# this is a function that Djurhuus references from https://rdrr.io/github/mworkentine/mattsUtils/src/R/microbiome_helpers.R#sym-add_taxonomy_column
split_species = function(string, n = 2) {
  splits = str_split(string, " ", n + 1)
  res = map_if(splits, ~length(.x) > 2, ~.x[1:n]) %>%
    map_chr(str_c, collapse = " ")
  return(res)
}

# this is function that Djurhuus references from https://rdrr.io/github/mworkentine/mattsUtils/src/R/microbiome_helpers.R#sym-add_taxonomy_column
add_taxonomy_column = function(physeq, num_species = 1) {
  tax_df = as.data.frame(tax_table(physeq)@.Data) %>%
    rownames_to_column("OTU") %>%
    mutate(specie = split_species(specie, n = num_species)) %>%
    mutate(Taxonomy =
      case_when(
        is.na(class)  ~ str_c("", phyla),
        is.na(order)  ~ str_c("", class),
        is.na(family)  ~ str_c("", order),
        is.na(genus)   ~ str_c("", family),
        is.na(specie) ~ str_c("", genus),
        TRUE ~ str_c(genus, " ", specie)
      )
    )

  tax = as.matrix(tax_df[, -1])
  rownames(tax) = tax_df$OTU
  tax_table(physeq) = tax_table(tax)

  return(physeq)
}

# Custom tax_agglomerate so that the level at which to agglomerate becomes first tax column
# This is because otherwise, conflicting tax IDs at higher level, for the same tax ID at the level to agglomerate yield two (or n) agglomerations as output - but you really just want 1. 
# provided by GitHub user Adriaaula as a response to the issue opened by Anni Djurhuus here: https://github.com/joey711/phyloseq/issues/941
tax_glom_custom = function(physeq, tax.col){
  
  require(tidyverse)
  
  reorderedtax = data.frame(tax_table(physeq)) %>% 
    # Puts the selected column first with everything else aftwds
    dplyr::select(one_of(tax.col), everything()) %>% 
    as.matrix() %>%                                                    # My addition
    tax_table()
  
  tax_table(physeq) = reorderedtax
  
  return(tax_glom(physeq, taxrank = tax.col))
  
}



# eDNA index directly from Kelly et al. 2019
# math defined below
eDNA_index <- function(x) { #where x is a dataframe with taxa/OTUs/etc in rows, and samples in columns
  rowMax <- function(x){apply(x, MARGIN = 1, FUN = max)}
  temp <- sweep(x, MARGIN = 2, STATS = colSums(x), FUN = "/") #create proportion for each taxon/OTU within each sample
  sweep(temp, MARGIN = 1, STATS = rowMax(temp), FUN = "/")
}
```
$$eDNA_{ij} = \dfrac{\dfrac{Y_{ij}}{\Sigma_i Y_i}}{max_{j}(\dfrac{Y_{ij}}{\Sigma_i Y_i})}$$


# metadata pre-processing
```{r Pre-processing metadata tables, message=FALSE}
# 12S metadata file
S12_metadata <-  read_tsv("./12S_exports/metadata12S.tsv") %>%
  column_to_rownames(var = "sampleid")
S12_metadata$size_fraction <- paste(as.character(S12_metadata$size_fraction), 'µm')
S12_metadata <- S12_metadata %>% 
  mutate(disc_depth = paste(as.character(S12_metadata$depth), 'm')) %>% 
  mutate(timeofyear = paste(as.character(S12_metadata$year), S12_metadata$season))

S12_samples_names <- c("2014F-GULD04-100mL", "2014F-GULD04-1mL", "2014S-GULD04-250mL", "2016F-GULD4-100m-L", "2016F-GULD4-250m-L", "2016S-GULD3-40m-L", "2017F-GULD04-1mL", "2017S-GULD04-60mL", "2018F-SG23-1mL", "2018F-SG23-20mL", "2018F-SG23-250mL", "2018F-SG23-60mL", "2018S-GULD04-250mL", "2019S-GULD04-1mL", "2019S-GULD04-250mL")

# 16S metadata file
S16_metadata <-  read_tsv("./16S_exports/metadata.tsv") %>%
  column_to_rownames(var = "sampleid")
S16_metadata$size_fraction <- paste(as.character(S16_metadata$size_fraction), 'µm')
S16_metadata <- S16_metadata %>%
  mutate(disc_depth = paste(as.character(S16_metadata$depth), 'm')) %>% 
  mutate(timeofyear = paste(as.character(S16_metadata$year), S16_metadata$season))

S16_samples_names <- c("2014F-GULD04-100mL", "2014F-GULD04-100mS", "2014F-GULD04-1mL",   "2014F-GULD04-1mS",   
                       "2014F-GULD04-20mL", "2014F-GULD04-20mS","2014F-GULD04-250mL", "2014F-GULD04-250mS", 
                       "2014S-GULD04-100mL", "2014S-GULD04-100mL" , "2014S-GULD04-1mL", "2014S-GULD04-1mS" ,  
                       "2014S-GULD04-20mL" , "2014S-GULD04-20mS" , "2014S-GULD04-250mL","2014S-GULD04-250mS", 
                       "2016F-GULD4-100m-L",  "2016F-GULD4-100m-S" , "2016F-GULD4-1m-L"   , "2016F-GULD4-1m-S" ,  
                       "2016F-GULD4-20m-L"  , "2016F-GULD4-20m-S" ,  "2016F-GULD4-250m-L" , "2016F-GULD4-250m-S" , 
                       "2016S-GULD3-1m-L"   ,"2016S-GULD3-1m-S"  ,  "2016S-GULD3-20m-L"  , "2016S-GULD3-20m-S"  , 
                       "2016S-GULD3-40m-L" ,  "2016S-GULD3-40m-S"  ,"2016S-GULD3-80m-L","2016S-GULD3-80m-S"  , 
                       "2017F-GULD04-1mL"   , "2017F-GULD04-1mS"  ,  "2017F-GULD04-20mL"  ,"2017F-GULD04-20mS" ,  
                       "2017F-GULD04-250mL" , "2017F-GULD04-250mS" , "2017F-GULD04-60mL"  , "2017F-GULD04-60mS"  ,
                       "2017S-GULD04-1mL"  ,  "2017S-GULD04-1mS"   , "2017S-GULD04-20mL" ,  "2017S-GULD04-20mS" ,  
                       "2017S-GULD04-250mL" ,"2017S-GULD04-250mS"  ,"2017S-GULD04-60mL"  , "2017S-GULD04-60mS"  , 
                       "2018F-SG23-1mL"    ,  "2018F-SG23-1mS"     ,"2018F-SG23-20mL"    , "2018F-SG23-20mS"    , 
                       "2018F-SG23-250mL"   , "2018F-SG23-250mS"  ,  "2018F-SG23-60mL", "2018F-SG23-60mS"    , 
                       "2018S-GULD04-1mS"  ,  "2018S-GULD04-20mL" ,  "2018S-GULD04-20mS" ,  "2018S-GULD04-250mL" ,
                       "2018S-GULD04-250mS" , "2018S-GULD04-60mL"  , "2018S-GULD04-60mS" ,  "2019S-GULD04-1mS"  ,  
                       "2019S-GULD04-20mL"  ,"2019S-GULD04-20mS"  , "2019S-GULD04-250mL" , "2019S-GULD04-250mS",  
                       "2019S-GULD04-60mL",   "2019S-GULD04-60mS" )

# 18S metadata file
S18_metadata <-  read_tsv("./18S_exports/metadata.tsv") %>%
  column_to_rownames(var = "sampleid")
S18_metadata$size_fraction <- paste(as.character(S18_metadata$size_fraction), 'µm')
S18_metadata <- S18_metadata %>%
  mutate(disc_depth = paste(as.character(S18_metadata$depth), 'm')) %>% 
  mutate(timeofyear = paste(as.character(S18_metadata$year), S18_metadata$season))

```

# Taxonomic tables pre-processing
```{r Pre-processing taxonomic tables, message=FALSE}
S12_tax <- read_tsv("./12S_exports/taxa_FishCard/taxonomy.tsv") %>%
  remove_rownames %>%
  column_to_rownames(var="Feature ID") %>%
  as.matrix()

#changing the taxonomic level to cut-off their pre-fix
S16_tax <- read_tsv("./16S_exports/taxa/taxonomy.tsv") %>%
      mutate_at("domain", str_replace, "D_0__", "")%>%
      mutate_at("phyla", str_replace, "D_1__", "")%>%
      mutate_at("class", str_replace, "D_2__", "")%>%
      mutate_at("order", str_replace, "D_3__", "")%>%
      mutate_at("family", str_replace, "D_4__", "")%>%
      mutate_at("genus", str_replace, "D_5__", "")%>%
      mutate_at("specie", str_replace, "D_6__", "")%>% 
      remove_rownames %>%
      column_to_rownames(var="Feature ID") %>%
      as.matrix()

#changing the taxonomic level to cut-off their pre-fix
S18_tax <- read_tsv("./18S_exports/taxa_final/taxonomy.tsv") %>%
      mutate_at("domain", str_replace, "d__", "")%>%
      mutate_at("phyla", str_replace, "p__", "")%>%
      mutate_at("class", str_replace, "c__", "")%>%
      mutate_at("order", str_replace, "o__", "")%>%
      mutate_at("family", str_replace, "f__", "")%>%
      mutate_at("genus", str_replace, "g__", "")%>%
      mutate_at("specie", str_replace, "s__", "")%>% 
      remove_rownames %>%
      column_to_rownames(var="Feature ID") %>%
      as.matrix()
```


# Loading in the data to create 1 phyloseq object per marker region
```{r Loading in various data sets for each marker, further (minimal) pre-processing, & merging into 1 phyloseq object, message=FALSE, warning = FALSE}

# 12S marker
S12_ASVs <- read_qza("./12S_exports/dada2_output/dada2_table_filt.qza")[["data"]] %>%
  as.data.frame() 
names(S12_ASVs) <- S12_samples_names
S12_ASVs <- otu_table(S12_ASVs, taxa_are_rows = TRUE)

S12_tax <- tax_table(S12_tax)
S12_metadata <- sample_data(S12_metadata)

S12_tree <- read_qza("./12S_exports/asvs-tree.qza")[["data"]] %>%
  read_tree(taxa_are_rows = TRUE)


# 16S marker
S16_ASVs <- read_qza("./16S_exports/dada2_output/dada2_table_final.qza")[["data"]] %>%
  as.data.frame() 
names(S16_ASVs) <- S16_samples_names
S16_ASVs <- otu_table(S16_ASVs, taxa_are_rows = TRUE)

S16_tax <- tax_table(S16_tax)
S16_metadata <- sample_data(S16_metadata)
S16_tree <- read_qza("./16S_exports/asvs-tree.qza")[["data"]] %>%
  read_tree(taxa_are_rows = TRUE)


# 18S marker
S18_ASVs <- read_qza("./18S_exports/dada2_output/dada2_table_filt.qza")[["data"]] %>%
  as.data.frame() %>%
  otu_table(taxa_are_rows = TRUE)
S18_tax <- tax_table(S18_tax)
S18_metadata <- sample_data(S18_metadata)
S18_tree <- read_qza("./18S_exports/asvs-tree.qza")[["data"]] %>%
  read_tree(taxa_are_rows = TRUE)

# Merging data into phyloseq object per marker region
(S12_physeq <-  merge_phyloseq(S12_ASVs, S12_tax, S12_metadata, S12_tree))
(S16_physeq <-  merge_phyloseq(S16_ASVs, S16_tax, S16_metadata, S16_tree))
(S18_physeq <-  merge_phyloseq(S18_ASVs, S18_tax, S18_metadata, S18_tree))
```

```{r agglomerate by family, warning=FALSE}
(S12_fam <- tax_glom(S12_physeq, "family"))
(S16_fam <- tax_glom(S16_physeq, "family"))
(S18_fam <- tax_glom(S18_physeq, "family"))
```


```{r Subset only taxa with a family annotation, assign the family name to new column: Taxonomy, make this column the taxa_names(), warning = FALSE}
S12_fam <- add_taxonomy_column(S12_fam) %>% 
  subset_taxa(Taxonomy !="NA") %>% 
  tax_glom_custom("family")
taxa_names(S12_fam) <- paste(S12_fam@tax_table[,1])

S16_fam <- add_taxonomy_column(S16_fam) %>% 
  subset_taxa(Taxonomy !="NA") %>% 
  tax_glom_custom("family")
taxa_names(S16_fam) <- paste(S16_fam@tax_table[,1])

S18_fam <- add_taxonomy_column(S18_fam) %>% 
  subset_taxa(Taxonomy !="NA") %>% 
  tax_glom_custom("family")
taxa_names(S18_fam) <- paste(S18_fam@tax_table[,1])
```


```{r Merging phyloseq objects without trees}
# Merge 16S and 18S Gully data where we have data from both markers (as read counts, at the family level)
(Gully_1618_fam <- merge_phyloseq(otu_table(S16_fam,taxa_are_rows = T),
                                 otu_table(S18_fam,taxa_are_rows = T),
                                 tax_table(S16_fam),
                                 tax_table(S18_fam),
                                 sample_data(S18_fam)       # 18S samples because it has one less sample than 16S (from QC)
                                 ))

# Merge all Gully data for samples where we have 12S data (as read counts, at the family level)
(Gully_all_fam <- merge_phyloseq(otu_table(S12_fam,taxa_are_rows = T),
                                 otu_table(S16_fam,taxa_are_rows = T),
                                 otu_table(S18_fam,taxa_are_rows = T),
                                 tax_table(S12_fam),
                                 tax_table(S16_fam),
                                 tax_table(S18_fam),
                                 sample_data(S12_fam)
                                 ))

```



```{r from Kelly et al 2019: eDNA index formula implementation}
# computing eDNA index for 16S and 18S merged data over 69 samples
Gully_1618_eDNA_otu <- as.data.frame(otu_table(Gully_1618_fam)) %>%
  eDNA_index() %>% 
  otu_table(taxa_are_rows = TRUE)
# Prune out taxa who dont have any data in any sample
Gully_1618_eDNA_otu = prune_taxa(taxa_sums(Gully_1618_eDNA_otu)>0, Gully_1618_eDNA_otu)
# Re-create a standardized phyloseq object
(Gully_1618_eDNA_indexed <- merge_phyloseq(Gully_1618_eDNA_otu,
                                 tax_table(Gully_1618_fam),
                                 sample_data(Gully_1618_fam)
                                 ))

# computing eDNA index for all markers merged data over 15 samples
Gully_all_eDNA_otu <- as.data.frame(otu_table(Gully_all_fam)) %>%
  eDNA_index() %>% 
  otu_table(taxa_are_rows = TRUE)
# Prune out taxa who dont have any data in any sample
Gully_all_eDNA_otu = prune_taxa(taxa_sums(Gully_all_eDNA_otu)>0, Gully_all_eDNA_otu)
# Re-create a standardized phyloseq object
(Gully_all_eDNA_indexed <- merge_phyloseq(Gully_all_eDNA_otu,
                                 tax_table(Gully_all_fam),
                                 sample_data(Gully_all_fam)
                                 ))

# Exporting the eDNA index standardized dataframes of families as rows, by samples as columns
Gully_1618_eDNA_indexed %>%
  otu_table() %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("family") %>% 
  write_tsv("Gully_1618_eDNA_indexed.tsv")

Gully_all_eDNA_indexed %>%
  otu_table() %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("family") %>% 
  write_tsv("Gully_all_eDNA_indexed.tsv")
```
# Plotting nMDS from Euclidean distances of the whole standardized community
```{r}
# Calculating distances
S1618_ord_nmds_eucl <- phyloseq::ordinate(Gully_1618_eDNA_indexed, method = "NMDS", distance = "euclidean")
# Plotting
plot_ordination(Gully_1618_eDNA_indexed, S1618_ord_nmds_eucl, color="depth")
ggsave("nMDS_Eucl_1618S_eDNAindexed.png")

# Calculating distances
all_ord_nmds_eucl <- phyloseq::ordinate(Gully_all_eDNA_indexed, method = "NMDS", distance = "euclidean")
# Plotting
plot_ordination(Gully_all_eDNA_indexed, all_ord_nmds_eucl, color="depth")
ggsave("nMDS_Eucl_all_eDNAindexed.png")
```

